### PR TITLE
Dynamic Product file support

### LINKED ISSUE #56: Create product without carbon.product file for C5
Currently the carbon.product is required for the following goals: publish-product and generate-runtime. Improve the plugin so that it does not require this file.

### PR DESCRIPTION
**Pull request by DilanUA**
This feature resolves #56 and adds an improvement to carbon-feature-plugin in supporting a dynamically generated product file for the reference of publish-product and generate-runtime maven goals. A dynamically generated product file during build-time overcomes the issue of having to manually update a statically maintained product file with version-bumps for a release-build, thus contributing to a fully-automated release process.

### CODE REVIEW

File: carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/Configurations.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import javax.xml.bind.annotation.XmlRootElement;
+
+/**
+ * Bean to represent configurations.
+ *
+ * This is used write configurations data to carbon.product file.
+ *
+ * @since 3.1.0
+ */
+
```
Comment (imesh) [inline carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/Configurations.java]: Might be better to remove the additional blank line between the class comment and class definition.

File: carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/Configurations.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+@XmlRootElement
+public class Configurations {
+
+    private List<Plugin> plugins;
+
```
Comment (imesh) [inline carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/Configurations.java]: We may not need to keep blank lines between private variables.

File: carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/Configurations.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    public List<Property> getProperties() {
+        return properties;
+    }
+
+    @XmlElement(name = "property")
+    public void setProperties(List<Property> properties) {
+        this.properties = properties;
+    }
+
```
Comment (imesh) [inline carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/Configurations.java]: Shall we remove this blank line?

File: carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/Feature.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+/**
+ * Bean to represent an individual feature.
+ *
+ * This is used to
+ * [1] Capture build-time product file individual feature data from distribution pom file.
+ * [2] Write individual feature data to carbon.product file.
+ *
+ * @since 3.1.0
+ */
+
```
Comment (imesh) [inline carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/Feature.java]: Might be better to remove additional blank line between the class command and class definition.

File: carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/Plugin.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+ */
+
+@XmlRootElement
+public class Plugin {
+
+    private String id;
+
+    private Boolean autoStart;
+
+    private Integer startLevel;
```
Comment (imesh) [inline carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/Plugin.java]: Might be better to remove blank lines between private variables. Thanks!

File: carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/config/ProductConfig.java [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+
+    private List<Plugin> getDefaultPluginConfig() {
+        List<Plugin> plugins = new ArrayList<>();
+        // adding default plugins wth default autoStart and startLevel configurations
+        plugins.add(new Plugin("org.eclipse.core.runtime", true, 4));
+        plugins.add(new Plugin("org.eclipse.equinox.common", true, 2));
+        plugins.add(new Plugin("org.eclipse.equinox.ds", true, 2));
+        plugins.add(new Plugin("org.eclipse.equinox.p2.reconciler.dropins", true, 4));
+        plugins.add(new Plugin("org.eclipse.equinox.simpleconfigurator", true, 1));
+        plugins.add(new Plugin("org.eclipse.update.configurator", true, 1));
```
Comment (imesh) [inline carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/config/ProductConfig.java]: May be we can sort these lines by the start up order. WDYT?

File: carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/config/ProductFileConfig.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+    public ProductConfig getProductConfig() {
+        return productConfig;
+    }
+
+    @SuppressWarnings("unused")
+    public void setProductConfig(ProductConfig productConfig) {
+        this.productConfig = productConfig;
+    }
+
```
Comment (imesh) [inline carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/config/ProductFileConfig.java]: May be we can remove this blank line. Thanks!

File: carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/Configurations.java (hunk 1 above)
Comment (DilanUA) [reply carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/Configurations.java]: fixed by commit 44febf8eea7d14299bd940434d9d6abfef549a00

File: carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/Configurations.java (hunk 2 above)
Comment (DilanUA) [reply carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/Configurations.java]: fixed by commit 44febf8eea7d14299bd940434d9d6abfef549a00

File: carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/Configurations.java (hunk 3 above)
Comment (DilanUA) [reply carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/Configurations.java]: I prefer to keep one line break at the top of the class and the bottom, to make things more clear and uncluttered. But, off-course, we can remove those lines if majority stands on that.

File: carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/Feature.java (hunk 4 above)
Comment (DilanUA) [reply carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/Feature.java]: fixed by commit 44febf8eea7d14299bd940434d9d6abfef549a00

File: carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/Plugin.java (hunk 5 above)
Comment (DilanUA) [reply carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/Plugin.java]: fixed by commit 44febf8eea7d14299bd940434d9d6abfef549a00

File: carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/config/ProductConfig.java (hunk 6 above)
Comment (DilanUA) [reply carbon-feature-plugin/src/main/java/org/wso2/maven/p2/beans/product/config/ProductConfig.java]: fixed by commit 44febf8eea7d14299bd940434d9d6abfef549a00

**Review by imesh — APPROVED** [review]

**Review by ChanakaCooray — DISMISSED** [review]
@DilanUA It is always better to use the imperative style in commit message subjects (adding -> add), also the "resolves 56" part can go to the commit body and have a meaningful message as the subject.

**Comment by imesh** [comment]
Great work Dilan! As we discussed offline may be we can remove the blank lines before the last brace of class definitions. Thanks!

**Pull request merged**